src/nodes/flow/layers/additive_coupling_layer.jl

```diff
@@ -0,0 +1,294 @@
+## TODO: fix for non-unity partition dimensions
+
+export AdditiveCouplingLayer
+export getf, getflow, getdim
+export forward, forward!, backward, backward!
+export jacobian, inv_jacobian
+export det_jacobian, absdet_jacobian, logdet_jacobian, logabsdet_jacobian
+export detinv_jacobian, absdetinv_jacobian, logdetinv_jacobian, logabsdetinv_jacobian
+
+import Base: eltype
+
+@doc raw"""
+The additive coupling layer specifies an invertible function ``{\bf{y}} = g({\bf{x}})`` following the specific structure (for the mapping ``g: \mathbb{R}^2 \rightarrow \mathbb{R}^2``):
+
+```math
+    \begin{align}
+        y_1 &= x_1 \\
+        y_2 &= x_2 + f(x_1)
+    \end{align}
+```
+
+where ``f(\cdot)`` denotes an arbitrary function with mapping ``f: \mathbb{R} \rightarrow \mathbb{R}``. This function can be chosen arbitrarily complex. Non-linear functions (neural networks) are often chosen to model complex relationships. From the definition of the model, invertibility can be easily achieved as
+
+```math
+    \begin{align}
+        x_1 &= y_1 \\
+        x_2 &= y_2 - f(y_1)
+    \end{align}
+```
+
+The current implementation only allows for the mapping ``g: \mathbb{R}^2 \rightarrow \mathbb{R}^2``, although this layer can be generalized for arbitrary input dimensions.
+
+`AdditiveCouplingLayer(f <: AbstractCouplingFlow)` creates the layer structure with function `f`.
+
+### Example
+```julia
+f = PlanarFlow()
+layer = AdditiveCouplingLayer(f)
+```
+
+This layer structure has been introduced in:
+
+Dinh, Laurent, David Krueger, and Yoshua Bengio. "Nice: Non-linear independent components estimation." _arXiv preprint_ arXiv:1410.8516 (2014).
+"""
+struct AdditiveCouplingLayer{ T <: NTuple{N,AbstractCouplingFlow} where { N } } <: AbstractCouplingLayer
```

Comment:
Interesting syntax, didn't know about that. Since I didn't know its possible I'm interested does it affect performance or not. Should be fine but we can discuss.

Reply:
The reason for this particular syntax was that it resulted in type-stable code.